python/cuml/preprocessing/encoders.py

```diff
@@ -0,0 +1,356 @@
+        The reason is because sparse matrix are not fully supported by cupy
+        yet, causing incorrect values when computing one hot encodings.
+        See https://github.com/cupy/cupy/issues/3223
+    dtype : number type, default=np.float
+        Desired datatype of transform's output.
+    handle_unknown : {'error', 'ignore'}, default='error'
+        Whether to raise an error or ignore if an unknown categorical feature
+        is present during transform (default is to raise). When this parameter
+        is set to 'ignore' and an unknown category is encountered during
+        transform, the resulting one-hot encoded columns for this feature
+        will be all zeros. In the inverse transform, an unknown category
+        will be denoted as None.
+
+    Attributes
+    ----------
+    drop_idx_ : array of shape (n_features,)
+        ``drop_idx_[i]`` is the index in ``categories_[i]`` of the category to
+        be dropped for each feature. None if all the transformed features will
+        be retained.
+    """
+    def __init__(self, categories='auto', drop=None, sparse=True,
+                 dtype=np.float, handle_unknown='error'):
+        self.categories = categories
+        self.sparse = sparse
+        self.dtype = dtype
+        self.handle_unknown = handle_unknown
+        self.drop = drop
+        self._fitted = False
+        self.drop_idx_ = None
+        self._features = None
+        self._encoders = None
+        self.input_type = None
+        if sparse and np.dtype(dtype) not in ['f', 'd', 'F', 'D']:
+            raise ValueError('Only float32, float64, complex64 and complex128 '
+                             'are supported when using sparse')
+
+    def _validate_keywords(self):
+        if self.handle_unknown not in ('error', 'ignore'):
+            msg = ("handle_unknown should be either 'error' or 'ignore', "
+                   "got {0}.".format(self.handle_unknown))
+            raise ValueError(msg)
+        # If we have both dropped columns and ignored unknown
+        # values, there will be ambiguous cells. This creates difficulties
+        # in interpreting the model.
+        if self.drop is not None and self.handle_unknown != 'error':
+            raise ValueError(
+                "`handle_unknown` must be 'error' when the drop parameter is "
+                "specified, as both would create categories that are all "
+                "zero.")
+
+    def _check_is_fitted(self):
+        if not self._fitted:
+            msg = ("This OneHotEncoder instance is not fitted yet. Call 'fit' "
+                   "with appropriate arguments before using this estimator.")
+            raise NotFittedError(msg)
+
+    def _compute_drop_idx(self):
+        """Helper to compute indices to drop from category to drop"""
+        if self.drop is None:
+            return None
+        elif isinstance(self.drop, str) and self.drop == 'first':
```

Comment:
Good point!

Reply:
Actually this is required in case drop is the wrong datatype (which we check in the `else` close) for instance if a `DataFrame` is used it should end up in the else clause but instead it fails on `elif self.drop == 'first':` because DataFrame cannot be compared to `str`.

Reply:
Fair enough!